sidecar/integration/azure_intergration_test.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	port, err := strconv.ParseInt(mongoDBPort, 10, strconv.IntSize)
+	if err != nil {
+		t.Errorf("env var 'MONGODB_PORT' should be an integer!")
+	}
+
+	config := &app.Configuration{
+		SharedSecret: "secret",
+		BaseDir:      baseDir,
+		Context: &types.Context{
+			Name:          "testmodule",
+			EventID:       "1111111",
+			CorrelationID: "fish",
+			ParentEventID: "",
+		},
+		ServerPort: 8080,
+		AzureBlobProvider: &azurestorage.Config{
+			BlobAccountName: os.Getenv("AZURE_STORAGE_ACCOUNT_NAME"),
+			BlobAccountKey:  os.Getenv("AZURE_STORAGE_ACCOUNT_KEY"),
+			ContainerName:   "frank",
+		},
+		MongoDBMetaProvider: &mongodb.Config{
+			Name:       os.Getenv("MONGODB_NAME"),
+			Password:   os.Getenv("MONGODB_PASSWORD"),
+			Collection: os.Getenv("MONGODB_COLLECTION"),
+			Port:       int(port),
+		},
+		PrintConfig: false,
+		LogLevel:    "Debug",
+	}
+
+	// Create Module #1
+	module1, err := createModule(config)
+	if err != nil {
+		t.Error(err)
+	}
+	defer module1.Close() // This is to ensure cleanup
+
+	// Write an output image blob
+	blob1 := "img1.png"
+	blob1FilePath := path.Join(outDataDir, blob1)
+	writeOutputBlob(blob1FilePath)
+
+	// Write an output image blob
+	blob2 := "img2.png"
+	blob2FilePath := path.Join(outDataDir, blob2)
+	writeOutputBlob(blob2FilePath)
+
+	// Grab the length of the output directory
+	outFiles, err := ioutil.ReadDir(outDataDir)
+	if err != nil {
+		t.Errorf("error reading out dir '%+v'", err)
+	}
+	outLength := len(outFiles)
+
+	// Write an output metadata file
+	insight := []byte(`[{"key": "key2","value": "value2"}]`)
+	writeOutputBytes(insight, outMetaFilePath)
+
+	// Write an output event file
+	j := fmt.Sprintf(`[{"key":"eventType","value":"%s"},{"key":"files","value":"%s,%s"},{"key":"abc","value":"123"}]`, eventTypes[0], blob1, blob2)
+	outEvent := []byte(j)
+
+	writeOutputBytes(outEvent, outEventFilePath)
+
+	client := &http.Client{}
+
+	// Ready will attempt to sync the execution environment for this module - this should be empty
+	if err := executeRequest(client, config.SharedSecret, fmt.Sprintf("%v", config.ServerPort), "ready"); err != nil {
+		t.Errorf("error calling ready '%+v'", err)
+	}
+
+	// Done will commit the written files to external providers
+	if err := executeRequest(client, config.SharedSecret, fmt.Sprintf("%v", config.ServerPort), "done"); err != nil {
+		t.Errorf("error calling done '%+v'", err)
+	}
+
+	// Clear local module directories
+	module1.Close()
+
+	// Hydrate event
+	b, err := ioutil.ReadFile(inEventFilePath)
+	if err != nil {
+		t.Errorf("error reading event from disk '%+v'", err)
+	}
+	var inEvent common.Event
+	err = json.Unmarshal(b, &inEvent)
+	if err != nil {
+		t.Errorf("error unmarshalling event '%+v'", err)
+	}
+
+	// Create Module #2
+	config.Context.ParentEventID = config.Context.EventID
+	config.Context.EventID = inEvent.EventID
+	module2, err := createModule(config)
+	if err != nil {
+		t.Error(err)
+	}
+	defer module2.Close()
```

Comment:
Why is this module defer close and Line137 not defer?

Reply:
defer cleans up the module's temp directory. The 1st is to ensure it runs in the event of failure. The second is to clean up before the 2nd modules starts. This is an idempotent call.

Reply:
Makes sense